cmd/hegel/root_cmd.go

```diff
@@ -105,52 +120,110 @@ func NewRootCommand() (*RootCommand, error) {
 	return rootCmd, nil
 }
 
-// PreRun satisfies cobra.Command.PreRunE and unmarshalls. Its responsible for populating rc.Opts.
-func (rc *RootCommand) PreRun(*cobra.Command, []string) error {
-	if err := rc.Vpr.Unmarshal(&rc.Opts); err != nil {
+// PreRun satisfies cobra.Command.PreRunE and unmarshalls. Its responsible for populating c.Opts.
+func (c *RootCommand) PreRun(*cobra.Command, []string) error {
+	if err := c.vpr.Unmarshal(&c.Opts); err != nil {
 		return err
 	}
 
-	return rc.validateOpts()
+	return c.validateOpts()
 }
 
-// Run executes Hegel. Its temporarily unimplemented.
-func (rc *RootCommand) Run(*cobra.Command, []string) error {
-	return nil
+// Run executes Hegel.
+func (c *RootCommand) Run(cmd *cobra.Command, _ []string) error {
+	logger, err := log.Init("github.com/tinkerbell/hegel")
+	if err != nil {
+		return fmt.Errorf("initialize logger: %w", err)
+	}
+	defer logger.Close()
+	metrics.Init(logger)
+
+	logger.Package("main").With("opts", fmt.Sprintf("%+v", c.Opts)).Info("root command options")
+
+	ctx, otelShutdown := otelinit.InitOpenTelemetry(cmd.Context(), "hegel")
+	defer otelShutdown(ctx)
+
+	metrics.State.Set(metrics.Initializing)
+
+	hardwareClient, err := hardware.NewClient(c.Opts.Facility, c.Opts.GetDataModel())
+	if err != nil {
+		return fmt.Errorf("create client: %w", err)
+	}
+
+	grpcServer := grpcserver.NewServer(logger, hardwareClient)
+
+	ctx, cancel := signals.ContextCanceller(cmd.Context(), syscall.SIGINT, syscall.SIGTERM)
```

Comment:
shouldn't this be using `ctx` returned by `otelinit.InitOpenTelementry`?

Reply:
Resolved.